crates/ruma-events/src/sticker.rs

```diff
@@ -4,9 +4,67 @@
 
 use ruma_common::OwnedMxcUri;
 use ruma_macros::EventContent;
-use serde::{Deserialize, Serialize};
+use serde::{de, Deserialize, Serialize};
 
-use crate::room::ImageInfo;
+#[cfg(feature = "compat-encrypted-stickers")]
+use crate::room::EncryptedFile;
+use crate::room::{ImageInfo, MediaSource};
+
+/// The source of a sticker media file.
+#[derive(Clone, Debug, Serialize)]
+#[allow(clippy::exhaustive_enums)]
+pub enum StickerMediaSource {
+    /// The MXC URI to the unencrypted media file.
+    #[serde(rename = "url")]
+    Plain(OwnedMxcUri),
+
+    /// The encryption info of the encrypted media file.
+    #[cfg(feature = "compat-encrypted-stickers")]
+    #[serde(rename = "file")]
+    Encrypted(Box<EncryptedFile>),
+}
+
+// Custom implementation of `Deserialize`, because serde doesn't guarantee what variant will be
+// deserialized for "externally tagged"¹ enums where multiple "tag" fields exist.
+//
+// ¹ https://serde.rs/enum-representations.html
+impl<'de> Deserialize<'de> for StickerMediaSource {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct StickerMediaSourceJsonRepr {
+            url: Option<OwnedMxcUri>,
+            #[cfg(feature = "compat-encrypted-stickers")]
+            file: Option<Box<EncryptedFile>>,
+        }
+
+        match StickerMediaSourceJsonRepr::deserialize(deserializer)? {
+            #[cfg(feature = "compat-encrypted-stickers")]
+            StickerMediaSourceJsonRepr { url: None, file: None } => {
+                Err(de::Error::missing_field("url"))
+            }
+            #[cfg(not(feature = "compat-encrypted-stickers"))]
+            StickerMediaSourceJsonRepr { url: None } => Err(de::Error::missing_field("url")),
```

Comment:
You should be able to simplify this to a single entry:

```suggestion
            StickerMediaSourceJsonRepr {
                url: None,
                #[cfg(feature = "compat-encrypted-stickers")]
                file: None,
            } => {
                Err(de::Error::missing_field("url"))
            }
```